File: samples_c/s128_custom_tokenizer.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

/* Girdi: tek satir, basit JSON-benzeri metin (yalniz {, }, [, ], :, ,,
   tirnakli string "...", sayi, true/false/null iceriyor - cJSON DEGIL,
   ozgun basit bir tokenizer).
   Cikti: her token kendi satirinda, "TUR:deger" bicimiyle (TUR: BRACE,
   BRACKET, COLON, COMMA, STRING, NUMBER, BOOL, NULL). */

static const char *p;

static void emit_string(void) {
    p++; /* acilis tirnak */
    char buf[256];
    int len = 0;
    while (*p && *p != '"' && len < 255) {
        buf[len++] = *p++;
    }
    buf[len] = 0;
    if (*p == '"') p++;
    printf("STRING:%s\n", buf);
}

static void emit_number(void) {
    char buf[64];
    int len = 0;
    while ((isdigit((unsigned char)*p) || *p == '-' || *p == '.') && len < 63) {
        buf[len++] = *p++;
    }
    buf[len] = 0;
    printf("NUMBER:%s\n", buf);
}

int main(void) {
    char line[1024];
    if (!fgets(line, sizeof(line), stdin)) return 0;
    line[strcspn(line, "\r\n")] = 0;
    p = line;

    while (*p) {
        if (*p == ' ' || *p == '\t') {
            p++;
        } else if (*p == '{' || *p == '}') {
            printf("BRACE:%c\n", *p);
            p++;
        } else if (*p == '[' || *p == ']') {
            printf("BRACKET:%c\n", *p);
            p++;
        } else if (*p == ':') {
            printf("COLON::\n");
            p++;
        } else if (*p == ',') {
            printf("COMMA:,\n");
            p++;
        } else if (*p == '"') {
            emit_string();
        } else if (isdigit((unsigned char)*p) || *p == '-') {
            emit_number();
        } else if (strncmp(p, "true", 4) == 0) {
            printf("BOOL:true\n");
            p += 4;
        } else if (strncmp(p, "false", 5) == 0) {
            printf("BOOL:false\n");
            p += 5;
        } else if (strncmp(p, "null", 4) == 0) {
            printf("NULL:null\n");
            p += 4;
        } else {
            p++;
        }
    }

    return 0;
}

File: translations_rust_haiku/s128_custom_tokenizer.rs
use std::io;

fn emit_string(bytes: &[u8], pos: &mut usize) {
    *pos += 1; // skip opening quote
    let mut buf = String::new();
    while *pos < bytes.len() && bytes[*pos] as char != '"' && buf.len() < 255 {
        buf.push(bytes[*pos] as char);
        *pos += 1;
    }
    if *pos < bytes.len() && bytes[*pos] as char == '"' {
        *pos += 1;
    }
    println!("STRING:{}", buf);
}

fn emit_number(bytes: &[u8], pos: &mut usize) {
    let mut buf = String::new();
    while *pos < bytes.len() {
        let c = bytes[*pos] as char;
        if c.is_ascii_digit() || c == '-' || c == '.' {
            buf.push(c);
            *pos += 1;
        } else {
            break;
        }
    }
    println!("NUMBER:{}", buf);
}

fn main() {
    let mut line = String::new();
    io::stdin().read_line(&mut line).ok();
    line = line.trim_end_matches('\n').trim_end_matches('\r').to_string();

    let bytes = line.as_bytes();
    let mut pos = 0;

    while pos < bytes.len() {
        let c = bytes[pos] as char;
        if c == ' ' || c == '\t' {
            pos += 1;
        } else if c == '{' || c == '}' {
            println!("BRACE:{}", c);
            pos += 1;
        } else if c == '[' || c == ']' {
            println!("BRACKET:{}", c);
            pos += 1;
        } else if c == ':' {
            println!("COLON::");
            pos += 1;
        } else if c == ',' {
            println!("COMMA:,");
            pos += 1;
        } else if c == '"' {
            emit_string(&bytes, &mut pos);
        } else if c.is_ascii_digit() || c == '-' {
            emit_number(&bytes, &mut pos);
        } else if bytes[pos..].starts_with(b"true") {
            println!("BOOL:true");
            pos += 4;
        } else if bytes[pos..].starts_with(b"false") {
            println!("BOOL:false");
            pos += 5;
        } else if bytes[pos..].starts_with(b"null") {
            println!("NULL:null");
            pos += 4;
        } else {
            pos += 1;
        }
    }
}
